### ISSUE TITLE
Loss over a padded and masked sequence

### ISSUE THREAD

**Issue by giamic** [tensorflow/tensorflow#39146]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 19.10
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): v2.1.0-rc2-17-ge5bf8de
- Python version: 3.7

**Describe the current behavior**
The data I'm working on is a collection of sequences of different length. I have padded all the sequences to the same length and written an LSTM model that uses masks to ignore the padded part of the data.

The loss at each step is given by the sum of the losses at all points in the sequences divided by the total length of the sequences, original + padded part.

**Describe the expected behavior**
I expected the loss to be given by the sum of the losses at all points in the sequences divided by the total length **of the original sequences**, discarding the padded part.

Is this the intended behavior? And are the two fundamentally equivalent, from the point of view of the backprop algorithm?

**Standalone code to reproduce the issue**
```
import numpy as np
import tensorflow as tf
from tensorflow.keras import Model
from tensorflow.keras.layers import LSTM

# Config
N = 32
T = 10
n = 2
mask_value = -1.
tf.random.set_seed(1)
np.random.seed(1)

# Data creation
X = np.ones((N, T, n)) * mask_value
Y = np.ones((N, T, 1)) * mask_value
for i in range(N):
    l = np.random.randint(1, T)
    value = np.random.random([l, n])
    X[i, :l] = value
    Y[i, :l] = np.array([sum(v) > 0.5 * n for v in value])[:, None]


class MyModel(Model):
    def __init__(self, n, mask_value, *args, **kwargs):
        super().__init__(name='MyModel', *args, **kwargs)
        self.mask_value = mask_value
        self.n = n
        self.LSTM = LSTM(self.n, return_sequences=True, activation='linear')
        return

    def call(self, inputs, training=None, mask=None):
        mask = tf.cast(tf.reduce_sum(inputs - self.mask_value, axis=-1), tf.bool)
        x = self.LSTM(inputs, mask=mask)
        return x


model = MyModel(n, mask_value)
model.build(input_shape=(N, T, n))
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy'],
)
model.summary()

mask = 1 - tf.cast(tf.reduce_all(tf.equal(X, mask_value), axis=-1), tf.float32)
loss_unmasked = tf.reduce_mean(tf.keras.losses.binary_crossentropy(Y, model.predict(X)))
loss_masked_1 = tf.reduce_sum(tf.keras.losses.binary_crossentropy(Y, model.predict(X)) * mask) / tf.reduce_sum(mask)
loss_masked_2 = tf.reduce_sum(tf.keras.losses.binary_crossentropy(Y, model.predict(X)) * mask) / (N * T)
print(f"model.evaluate(X, Y): {model.evaluate(X, Y)[0]:.2f}\n"
      f"loss_unmasked       : {loss_unmasked:.2f}\n"
      f"loss_masked_1       : {loss_masked_1:.2f}\n"
      f"loss_masked_2       : {loss_masked_2:.2f}"
      )
```

**Comment by amahendrakar** [comment]
Got similar values with [TF v2.1](https://colab.research.google.com/gist/amahendrakar/f82d3f6099c8bd27fbe75e48b32d4c65/39146-2-1.ipynb) and [TF v2.2.0-rc4](https://colab.research.google.com/gist/amahendrakar/ea34587561bf48b61cea45314354f330/39146-2-2.ipynb). Please find the attached gist. Thanks!

**Comment by giamic** [comment]
Yes, this is the behavior I see as well: the loss given by the model is equal to `loss_masked_2`. However, I believe it should be equal to `loss_masked_1`. Let me explain why this might be problematic.

For example, let's consider two identical data points. With the current implementation their contribution to the loss changes depending on the batch they find themselves in. Let us say that the first one is in a batch which has been padded to length `N` and contributes to the total loss of a quantity `L`; if the second one is padded to length `2N`, its contribution to the loss will be `L / 2`. This seems contradictory with the whole idea of masking.

After delving a bit more into the training loops, I suspect that it doesn't impact the training anyway because the value shown is actually never used. However, it reports incorrect information to the user.

**Comment by saikumarchalla** [comment]
Was able to reproduce the issue in TF 2.5 and Nightly versions.Please find the gist [here](https://colab.research.google.com/gist/saikumarchalla/f31da23b9d66dac0f32d003e7d9d9b9d/untitled79.ipynb).

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
